README.rst

===============================================
ElectrumX - Reimplementation of electrum-server
===============================================

This version of ElectrumX has been slightly modified to support blocks with any order of transactions. It is compatible with both the Bitcoin ABC canonical transaction ordering that will activate in November 2018, and the traditional topological transaction ordering used in all other bitcoin consensus rules. Your ElectrumX server will identify as version **ElectronX**. Please submit bug reports relating to transaction ordering to `our repo <https://github.com/Electron-Cash/electrumx>`_, otherwise send bug reports (and donations) to upstream author `Neil Booth <https://github.com/kyuupichan/electrumx>`_.

For a future network with bigger blocks.

  :Licence: MIT
  :Language: Python (>= 3.6)
  :Author: Neil Booth

Documentation
=============

See `readthedocs <https://electrumx.readthedocs.io/>`_.


**Neil Booth**  kyuupichan@gmail.com  https://github.com/kyuupichan

1BWwXJH3q6PRsizBkSGm2Uw4Sz1urZ5sCj


docs/changelog.rst
===========
 ChangeLog
===========

.. note:: It is strongly recommended you upgrade to Python 3.7, which
   fixes bugs in asyncio that caused an ever-growing open file count
   and memory consumption whilst serving clients.  Those problems
   should not occur with Python 3.7.

.. note:: The primary difference between this and the official ElectrumX
   is that this version supports any block ordering (including the lexically-
   ordered blocks that can be created/followed by several Bitcoin Cash node
   implementations: ABC, BU, XT, bchd, bcash, Bitprim, and others.). We also
   employ a different anti-phishing strategy than they do, based on a
   community-maintained blacklist.json file as well as manual banning
   to supplement that via the banip command. We also have slightly more
   generous limits to things like MAX_SEND and MAX_SUBS than the defaults
   of ElectrumX.  We also don't implement their newest cost control logic which
   we feel is too restrictive for normal BCH usage of the server.


Version 1.10.1 (21 April 2019)
==============================

* Added PEER_DISCOVERY_TOR environment variable (default off). If on, then
  ElectronX will behave as before this version and always forward/discover
  *.onion peers.  If off, then it will completely ignore *.onion peers and
  never forward them.  This policy change has been implemented because of the
  ease with which sybil attacks are possible if PEER_DISCOVER_TOR is enabled.
  (cculianu)
* Added more stuff to `getenv` rpc command display (cculinau)
* Optimized the ban handling a little bit to refuse peers earlier in the
  lifecycle of peer_discovery if a host is banned. (cculianu)

Version 1.10.0 (14 April 2019)
==============================

* Added banip, unbanip, banhost, unbanhost, listbanned, getenv, and
  session_ip_counts rpc commands (cculianu)
* Disallow multiple servers from same IP to appear in peers (doesn't apply
  to Tor, however) (cculianu)
* Limit client connections per IP to 50 by default (MAX_SESSIONS_PER_IP)
  to prevent the "clients exhausting FDs on server" attack vector (doesn't
  apply to Tor or localhost clients).
  (cculianu)
* Limit client connections per IP for Tor to 1000 by default (MAX_SESSIONS_TOR)
  Note that this limit also applies to clients coming from localhost even if
  not using Tor. (cculianu)
* Added blacklist.json mechanism for downloading a community maintained
  list of bad peers as an anti-phishing, anti-sybil attack measure.
  Use BLACKLIST_URL="" in env to disable, but it is recommended you leave
  it enabled.  We use our own file format for this but the code for this
  subsystem also understand the ElectrumX 1.10.0 format as well. (cculianu)
* Set MAX_SEND to 4000000 (4MB) by default (was originally 1MB which fails
  on some BCH wallets). (cculianu)
* Raised server limit from 250,000 subs to 1,000,000 subs by default
  (corresponds to env var MAX_SUBS) (cculianu)
* Raised default BANDWIDTH_LIMIT from 2MB to 8MB (8000000 bytes).  This is
  because very busy BCH wallets with huge histories would need a higher limit.
  Note the limit is a soft limit in bytes per hour, and it is not catastrophic
  to send a client 8MB in 1 hour. (cculianu)


Version 1.9.4 (7 Feb 2019)
============================
* Synch up to ElectrumX codebase (pulling in relevant & useful commits)
  (cculianu)
* coin additions / updates: SmartCash (rc125), NIX (phamels), Minexcoin
  (joesixpack), BitcoinCash (mblunderburg), Dash (zebra-lucky),
  BitcoinCashRegtest (ezegom), AXE (slowdive), NOR (flo071),
  BitcoinPlus (bushsolo), Myriadcoin (cryptapus), Trezarcoin (ChekaZ), Bitcoin
  Diamond (John Shine), NMC (jeremyrand), Dash (zebra-lucky), PeerCoin
  (peerchemist), BCH testnet (Mark Lundeberg, cculianu), Unitus (ChekaZ)
* close ElectrumX issues: `#554`_, `#653`_, `#655`_, `#684`_, `#713`_
* other minor tweaks (Michael Schmoock, Michael Taborsky)
* increase default MAX_SEND for AuxPow Chains.  Truncate AuxPow for block
  heights covered by a checkpoint.  (jeremyrand)
* tighter RPC param checking (ghost43)
* require aiorpcX 0.10.3

Version 1.8.12 (10 Nov 2018)
============================

* bug fix

Version 1.8.11 (07 Nov 2018)
============================

* require aiorpcX 0.10.1

Version 1.8.10 (05 Nov 2018)
============================

* require aiorpcX 0.10.0
* fix `#632`_
* coin additions / updates: ZelCash (TheTrunk)

Version 1.8.9 (02 Nov 2018)
===========================

* fix `#630`_

Version 1.8.8 (01 Nov 2018)
===========================

* require aiorpcX 0.9.0
* coin additions / updates: decred (dajohi, bolapara), zcash (erasmospunk),
  namecoin (JeremyRand),CivX (turcol), NewYorkCoin (erasmospunk)
* fix `#603`_, `#608`_
* other minor fixes and changes: FMCorz

Version 1.8.7 (13 Sep 2018)
===========================

* require aiorpcX 0.8.1
* fix reorg bug loading blocks from disk (erasmospunk)

Version 1.8.6 (12 Sep 2018)
===========================

* require aiorpcX 0.8.0
* suppress socket.send() errors
* new coin TokenPay (samfiragabriel)
* minor fix: wakiyamap

Version 1.8.5 (18 Aug 2018)
===========================

* require aiorpcX 0.7.3 which contains a couple of bugfixes
* fix `#552`_, `#577`_
* fixed a session limiting bug reported by ghost43
* coin additions / updates: PIVX and Decred Testnets, BitcoinGreen (cunhasb)
  Monacoin (wakayamap)
* proper generation input handling for various altcoins (erasmospunk) fixing
  `#570`_

Version 1.8.4 (14 Aug 2018)
===========================

* improved notification handling and efficiency
* improved daemon handling with minor fixes; full tests for Daemon class
* remove chain_state class
* various internal cleanups and improvements (erasmospunk)
* add PIVX support (erasmospunk) - mempool handling WIP
* fix protocol 1.3 handling of blockchain.block.header RPC (ghost43)

Version 1.8.3 (11 Aug 2018)
===========================

* separate the DB and the BlockProcessor objects
* comprehensive mempool tests
* fix `#521`_, `#565`_, `#567`_


electrumx/server/env.py
# Copyright (c) 2016, Neil Booth
#
# All rights reserved.
#
# See the file "LICENCE" for information about the copyright
# and warranty status of this software.

'''Class for handling environment configuration and defaults.'''


import re
import resource
from collections import namedtuple
from ipaddress import ip_address

from electrumx.lib.coins import Coin
from electrumx.lib.env_base import EnvBase
import electrumx.lib.util as lib_util


NetIdentity = namedtuple('NetIdentity', 'host tcp_port ssl_port nick_suffix')


class Env(EnvBase):
    '''Wraps environment configuration. Optionally, accepts a Coin class
       as first argument to have ElectrumX serve custom coins not part of
       the standard distribution.
    '''

    # Peer discovery
    PD_OFF, PD_SELF, PD_ON = range(3)

    def __init__(self, coin=None):
        super().__init__()
        self.obsolete(['UTXO_MB', 'HIST_MB', 'NETWORK'])
        self.db_dir = self.required('DB_DIRECTORY')
        self.db_engine = self.default('DB_ENGINE', 'leveldb')
        self.daemon_url = self.required('DAEMON_URL')
        if coin is not None:
            assert issubclass(coin, Coin)
            self.coin = coin
        else:
            coin_name = self.required('COIN').strip()
            network = self.default('NET', 'mainnet').strip()
            self.coin = Coin.lookup_coin_class(coin_name, network)
        self.cache_MB = self.integer('CACHE_MB', 1200)
        self.host = self.default('HOST', 'localhost')
        self.reorg_limit = self.integer('REORG_LIMIT', self.coin.REORG_LIMIT)
        # Server stuff
        self.tcp_port = self.integer('TCP_PORT', None)
        self.ssl_port = self.integer('SSL_PORT', None)
        if self.ssl_port:
            self.ssl_certfile = self.required('SSL_CERTFILE')
            self.ssl_keyfile = self.required('SSL_KEYFILE')
        self.rpc_port = self.integer('RPC_PORT', 8000)
        self.banner_file = self.default('BANNER_FILE', None)
        self.tor_banner_file = self.default('TOR_BANNER_FILE',
                                            self.banner_file)
        self.anon_logs = self.boolean('ANON_LOGS', False)
        self.log_sessions = self.integer('LOG_SESSIONS', 3600)
        # Peer discovery
        self.peer_discovery = self.peer_discovery_enum()
        self.peer_announce = self.boolean('PEER_ANNOUNCE', True)
        self.force_proxy = self.boolean('FORCE_PROXY', False)
        self.tor_proxy_host = self.default('TOR_PROXY_HOST', 'localhost')
        self.tor_proxy_port = self.integer('TOR_PROXY_PORT', None)
        self.peer_discovery_tor = self.peer_discovery_tor_parse()
        # The electrum client takes the empty string as unspecified
        self.donation_address = self.default('DONATION_ADDRESS', '')
        # Server limits to help prevent DoS
        self.max_send = self.integer('MAX_SEND', self.coin.DEFAULT_MAX_SEND)
        self.max_subs = self.integer('MAX_SUBS', 1000000)
        self.max_sessions = self.sane_max_sessions()
        self.max_sessions_per_ip = self.integer('MAX_SESSIONS_PER_IP', 50)
        self.max_sessions_tor = max(self.integer('MAX_SESSIONS_TOR', 1000), 10)  # Require at least 10 for MAX_SESSIONS_TOR
        self.max_session_subs = self.integer('MAX_SESSION_SUBS', 50000)
        self.bandwidth_limit = self.integer('BANDWIDTH_LIMIT', 8000000)
        self.session_timeout = self.integer('SESSION_TIMEOUT', 600)
        self.drop_client = self.custom("DROP_CLIENT", None, re.compile)
        # Blacklist URL. Set this to the empty string in the environment if you want to disable this facility
        self.blacklist_url = self.default('BLACKLIST_URL',
                                          'https://raw.githubusercontent.com/Electron-Cash/electronx-blacklist/master/blacklist.json'
                                          )
        self.blacklist_poll_interval = self.custom('BLACKLIST_POLL_INTERVAL', 300,
                                                   # parse as integer and limit it to >= 30 secs and <= 86400 (1 day)
                                                   lambda x: (int(x) >= 30 and int(x) <= 86400 and int(x)) or 300)
        # If this is set, then we ban whenever an IP address has more than self.max_sessions_per_ip connections
        self.ban_excessive_connections = self.boolean('BAN_EXCESSIVE_CONNECTIONS', True)

        # Identities
        clearnet_identity = self.clearnet_identity()
        tor_identity = self.tor_identity(clearnet_identity)
        self.identities = [identity
                           for identity in (clearnet_identity, tor_identity)
                           if identity is not None]

    def sane_max_sessions(self):
        '''Return the maximum number of sessions to permit.  Normally this
        is MAX_SESSIONS.  However, to prevent open file exhaustion, ajdust
        downwards if running with a small open file rlimit.'''
        env_value = self.integer('MAX_SESSIONS', 1000)
        nofile_limit = resource.getrlimit(resource.RLIMIT_NOFILE)[0]
        # We give the DB 250 files; allow ElectrumX 100 for itself
        value = max(0, min(env_value, nofile_limit - 350))
        if value < env_value:
            self.logger.warning('lowered maximum sessions from {:,d} to {:,d} '
                                'because your open file limit is {:,d}'
                                .format(env_value, value, nofile_limit))
        return value

    def clearnet_identity(self):
        host = self.default('REPORT_HOST', None)
        if host is None:
            return None
        try:
            ip = ip_address(host)
        except ValueError:
            bad = (not lib_util.is_valid_hostname(host)
                   or host.lower() == 'localhost')
        else:
            bad = (ip.is_multicast or ip.is_unspecified
                   or (ip.is_private and self.peer_announce))
        if bad:
            raise self.Error('"{}" is not a valid REPORT_HOST'.format(host))
        tcp_port = self.integer('REPORT_TCP_PORT', self.tcp_port) or None
        ssl_port = self.integer('REPORT_SSL_PORT', self.ssl_port) or None
        if tcp_port == ssl_port:
            raise self.Error('REPORT_TCP_PORT and REPORT_SSL_PORT '
                             'both resolve to {}'.format(tcp_port))
        return NetIdentity(
            host,
            tcp_port,
            ssl_port,
            ''
        )

    def tor_identity(self, clearnet):
        host = self.default('REPORT_HOST_TOR', None)
        if host is None:
            return None
        if not host.endswith('.onion'):
            raise self.Error('tor host "{}" must end with ".onion"'
                             .format(host))

        def port(port_kind):
            '''Returns the clearnet identity port, if any and not zero,
            otherwise the listening port.'''
            result = 0
            if clearnet:
                result = getattr(clearnet, port_kind)
            return result or getattr(self, port_kind)

        tcp_port = self.integer('REPORT_TCP_PORT_TOR',
                                port('tcp_port')) or None
        ssl_port = self.integer('REPORT_SSL_PORT_TOR',
                                port('ssl_port')) or None
        if tcp_port == ssl_port:
            raise self.Error('REPORT_TCP_PORT_TOR and REPORT_SSL_PORT_TOR '
                             'both resolve to {}'.format(tcp_port))

        return NetIdentity(
            host,
            tcp_port,
            ssl_port,
            '_tor',
        )

    def hosts_dict(self):
        return {identity.host: {'tcp_port': identity.tcp_port,
                                'ssl_port': identity.ssl_port}
                for identity in self.identities}

    def peer_discovery_enum(self):
        pd = self.default('PEER_DISCOVERY', 'on').strip().lower()
        if pd in ('off', ''):
            return self.PD_OFF
        elif pd == 'self':
            return self.PD_SELF
        else:
            return self.PD_ON

    def peer_discovery_tor_parse(self):
        tpd = self.default('PEER_DISCOVERY_TOR', 'off').strip().lower()  # if 'on', we forward .onion peers we hear about. otherwise off (due to sybil attack issues)
        if tpd in ('on', 'yes', 'true', '1', 'enabled', 'enable'):
            return True
        return False


    def get_info(self):
        ''' Used by rpc_getenv to get all the env vars we are using. '''
        env = {}
        for name in dir(self).copy():
            if name.startswith('_'):
                continue
            value = getattr(self, name, None)
            if not isinstance(value, (str, int, float, bool)):
                continue
            if isinstance(value, bool):
                value = "1" if value else ""  # Map bools to non-empty/empty strings to match how they are parsed
            env[name.upper()] = value
        # now mogrify some values
        env['PEER_DISCOVERY'] = {
            self.PD_ON : "on", self.PD_OFF : "off", self.PD_SELF : "self"
        }.get(self.peer_discovery, 'on')
        env['PEER_DISCOVERY_TOR'] = "on" if self.peer_discovery_tor else "off"
        clear_id = self.clearnet_identity()
        tor_id = self.tor_identity(clear_id)
        if tor_id is not None:
            env['REPORT_HOST_TOR'] = str(tor_id.host)
            env['REPORT_TCP_PORT_TOR'] = str(tor_id.tcp_port)
            env['REPORT_SSL_PORT_TOR'] = str(tor_id.ssl_port)
        if clear_id is not None:
            env['REPORT_HOST'] = str(clear_id.host)
            env['REPORT_TCP_PORT'] = str(clear_id.tcp_port)
            env['REPORT_SSL_PORT'] = str(clear_id.ssl_port)
        return env
